daemon/container_operations.go

```diff
@@ -574,52 +574,86 @@ func hasUserDefinedIPAddress(ipamConfig *networktypes.EndpointIPAMConfig) bool {
 	return ipamConfig != nil && (len(ipamConfig.IPv4Address) > 0 || len(ipamConfig.IPv6Address) > 0)
 }
 
-// User specified ip address is acceptable only for networks with user specified subnets.
-func validateNetworkingConfig(n *libnetwork.Network, epConfig *networktypes.EndpointSettings) error {
-	if n == nil || epConfig == nil {
+// validateEndpointSettings checks whether the given epConfig is valid. The nw parameter might be nil as a container
+// can be created with a reference to a network that don't exist yet. In that case, only partial validation will be
+// done.
+func validateEndpointSettings(nw *libnetwork.Network, nwName string, epConfig *networktypes.EndpointSettings) error {
+	if epConfig == nil {
 		return nil
 	}
-	if !containertypes.NetworkMode(n.Name()).IsUserDefined() {
+
+	var errs []error
+	if !containertypes.NetworkMode(nwName).IsUserDefined() {
+		// On Linux, user specified IP address is accepted only for networks with user specified subnets.
 		if hasUserDefinedIPAddress(epConfig.IPAMConfig) && !enableIPOnPredefinedNetwork() {
```

Comment:
Just an observation here; 

-  `hasUserDefinedIPAddress()` will always be `false` if there's no ipamConfig, which _may_ already invalidate other checks.
- ^^ slightly leaning toward inlining that function here, and just make it a `hasUserDefinedIPAddress` variable that we set (and use), as that function is _only_ used inside this check, and we're now potentially calling it twice with the same results
- _BOTH_ `enableIPOnPredefinedNetwork` and `serviceDiscoveryOnDefaultNetwork` are effectively `if runtime.GOOS == "windows"`)
- ^^ wondering if we need to split the _validation_ there, instead of having two utility functions that effectively means "different validation on windows vs linux" (not sure)

Reply:
Hm, most of these would be better handled in their own PR. Also, the next step once we get proper support for multi `--network` flags, is to make the default bridge just a "normal" bridge (ie. no more links, no more artificial restriction, etc...). So better keep these refactorings for that time (unless you want to go ahead and do this beforehand).

Reply:
> * BOTH `enableIPOnPredefinedNetwork` and `serviceDiscoveryOnDefaultNetwork` are effectively if runtime.GOOS == "windows")
> * ^^ wondering if we need to split the validation there, instead of having two utility functions that effectively means "different validation on windows vs linux" (not sure)

I added a `TODO` to remind us we should remove these functions once we remove legacy networking mode from the default bridge. As such, I think we shouldn't make the validation OS-dependent here, as it's going to be addressed (hopefully) soon.